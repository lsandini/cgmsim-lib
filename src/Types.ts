import { TypeDateISO } from './TypeDateISO';
import { drugs } from './drug';

/**
 * Represents the direction of blood glucose change.
 */
export type Direction =
	| 'DoubleDown'
	| 'SingleDown'
	| 'FortyFiveDown'
	| 'Flat'
	| 'FortyFiveUp'
	| 'SingleUp'
	| 'DoubleUp';

/**
 * Represents drug information.
 */
export type Drug = {
	/** The timestamp of the drug entry. */
	time: number;
	/** The name of the drug. */
	drug: string;
	/** Additional notes for the drug. */
	notes: string;
	/** The amount of insulin associated with the drug. */
	insulin: number;
	/** Placeholder property for empty space. */
	empty_space: any;
};

/**
 * Represents a note with associated notes.
 */
export type Note = {
	/** The type of note. */
	type: 'Note';
	/** The content of the note. */
	notes: string;
};

/**
 * Represents data related to physical activity.
 */
export type Activity = {
	/** The number of steps taken (optional). */
	steps?: number;
	/** The heart rate during the activity (optional). */
	heartRate?: number;
	/** The timestamp indicating when the activity data was recorded in ISO format. */
	created_at: TypeDateISO;
};

/**
 * Represents a blood glucose entry.
 */
export type Sgv = {
	/** The timestamp of the blood glucose entry in milliseconds. */
	mills: number;
	/** The blood glucose value. */
	sgv: number;
};

/**
 * Represents a blood glucose entry value type.
 */
export type EntryValueType = {
	/** The blood glucose value. */
	sgv: number;
	/** The direction of blood glucose change. */
	direction: string;
};

/**
 * Represents a blood glucose entry with additional date information.
 */
export type Entry = EntryValueType & {
	/** The timestamp of the entry. */
	date: number;
	/** The date in string format. */
	dateString: string;
	/** The type of entry. */
	type: 'sgv';
};

/**
 * Represents parameters for a profile.
 */
export type ProfileParams = {
	/** The basal insulin rate. */
	basal: number | { value: number; time: string; timeAsSecond?: number }[];
};

/**
 * Represents a profile with associated profile parameters.
 */
export type NSProfile = {
	/** The start date of the profile. */
	startDate: string;
	/** The default profile name. */
	defaultProfile: string;
	/** The store of profile parameters. */
	store: {
		[profileName: string]: ProfileParams;
	};
};

//Note', 'Profile Switch', 'Temp Basal', 'Meal Bolus', 'Bolus Wizard', 'Correction Bolus', 'Temporary Target', 'Insulin Change', 'Site Change', 'Sensor Start', 'Sensor Stop'

/**
 * Represents the treatment information for a Meal Bolus event.
 */
export type MealBolusTreatment = {
	/** The type of event, set to 'Meal Bolus'. */
	eventType:
		| 'Meal Bolus'
		| 'Bolus'
		| 'Correction Bolus'
		| 'Bolus Wizard'
		| 'Carb Correction';
	/** The amount of insulin administered for the meal bolus. */
	insulin?: number;
	/** The number of carbohydrates consumed. */
	carbs?: number;
	/** The date of the treatment creation in ISO format. */
	created_at: TypeDateISO;
};

/**
 * Represents the treatment information for a Profile Switch event.
 */
export type ProfileSwitchTreatment = {
	/** The type of event, set to 'Profile Switch'. */
	eventType: 'Profile Switch';
	/** The duration of the profile switch in minutes. */
	duration: number;
	/** The date of the treatment creation in ISO format. */
	created_at: TypeDateISO;
	/** The JSON representation of the new profile. */
	profileJson: string;
	/** The percentage change applied during the profile switch. */
	percentage: number;
};

/**
 * Represents the treatment information for a Temporary Basal event.
 */
export type TempBasalTreatment = {
	/** The type of event, set to 'Temp Basal'. */
	eventType: 'Temp Basal';
	/** The absolute insulin rate for the temporary basal. */
	absolute: number;
	/** The duration of the temporary basal in minutes. */
	duration: number;
	/** The date of the treatment creation in ISO format. */
	created_at: TypeDateISO;
};

/**
 * Represents the announcement as treatment.
 */
export type AnnouncementTreatment = {
	/** The date of the treatment creation in ISO format. */
	created_at: TypeDateISO;
	/** The type of event, set to 'Announcement'. */
	eventType: 'Announcement';
	/** Additional notes for the treatment. */
	notes: string;
};

/**
 * Represents treatment data.
 */
export type NSTreatment =
	| MealBolusTreatment
	| ProfileSwitchTreatment
	| TempBasalTreatment
	| AnnouncementTreatment;

export const isMealBolusTreatment = (
	treatment: NSTreatment,
): treatment is MealBolusTreatment =>
	treatment.eventType === 'Meal Bolus' ||
	treatment.eventType === 'Bolus' ||
	treatment.eventType === 'Bolus Wizard' ||
	treatment.eventType === 'Correction Bolus';

export const isAnnouncementTreatment = (
	treatment: NSTreatment,
): treatment is AnnouncementTreatment => treatment.eventType === 'Announcement';

export type NSTreatmentParsed = {
	drug: string;
	units: number;
	minutesAgo: number;
};

/**
 * Represents a treatment delta with additional minutes ago information.
 */
export type TreatmentBiexpParam = {
	units: number;
	minutesAgo: number;
	duration: number;
	peak: number;
};

/**
 * Represents a gender type.
 */
export type GenderType = 'Male' | 'Female';

/**
 * Represents parameters for a CGM simulation.
 */
export type CGMSimParams = {
	basalActivity: number;
	liverActivity: number;
	carbsActivity: number;
	bolusActivity: number;
	cortisoneActivity: number;
	alcoholActivity: number;
};

/**
 * Represents user parameters.
 */
export type UserParams = {
	nsUrl: string;
};

type PatientInfoBase = {
	/** Weight of the simulated user (in cm). */
	WEIGHT: number;
	/** Age of the simulated user. */
	AGE: number;
	/** Gender of the simulated user ('Male' or 'Female'). */
	GENDER: GenderType;
};

/**
 * Represents patient parameters for a cgmsim simulation.
 */
export type PatientInfoUva = PatientInfoBase;

/**
 * Represents patient parameters for a cgmsim simulation.
 */

export type PatientInfoCgmsim = PatientInfoBase & {
	/** Carbohydrate Ratio (CR) for insulin calculation. */
	CR: number;
	/** Insulin Sensitivity Factor (ISF) for insulin calculation. */
	ISF: number;
	/** Time taken for carbohydrates to be absorbed (in minutes default 360). */
	CARBS_ABS_TIME: number;
	/** Time period for insulin activity (Time Peak) (in minutes default 75). */
	TP: number;
	/** Duration of Insulin Action (DIA) (in hours default: 6). */
	DIA: number;
};

type MainParamsBase = {
	patient: PatientInfoBase;
	treatments: NSTreatment[];
	profiles: NSProfile[];
	pumpEnabled: boolean;
	activities?: Activity[];
	user: UserParams;
};
/**
 * Represents main parameters for a UVA simulation.
 */
export type MainParamsUVA = MainParamsBase & {
	lastState: UvaPatientState;
};

/**
 * Represents main parameters for a simulation.
 */
export type MainParams = MainParamsBase & {
	patient: PatientInfoCgmsim;
	entries: Sgv[];
};

/**
 * Represents the result of a simulation.
 */
export type SimulationResult = {
	sgv: number;
	deltaMinutes: number;
	carbsActivity: number;
	cortisoneActivity: number;
	basalActivity: number;
	bolusActivity: number;
	liverActivity: number;
	activityFactor: number;
	alcoholActivity: number;
	isf: { dynamic: number; constant: number };
};

/**
 * Represents the output of a UVA simulation.
 */
export type UvaOutput = {
	/** Glucose in plasma. */
	Gp: number;
	/** Glucose. */
	G?: number;
};

/**
 * Represents user parameters for a UVA simulation.
 */
export type UvaUserParams = {
	/** Insulin infusion rate. */
	iir: number;
	/** Insulin bolus amount. */
	ibolus: number;
	/** Carbohydrate intake. */
	carbs: number;
	/** Intensity of the simulation. */
	intensity: number;
};

/**
 * Represents a time delta for a UVA simulation.
 */
export type UvaDelta = 1;

/**
 * Represents an interval for a UVA simulation.
 */
export type UvaInterval = 5;

/**
 * Represents the state of a UVA patient.
 */
export type UvaPatientState = {
	/** Glucose in plasma. */
	Gp: number;
	/** Glucose in tissue. */
	Gt: number;
	/** Insulin delay compartment 1. */
	I_: number;
	/** Insulin in liver. */
	Il: number;
	/** Insulin in plasma. */
	Ip: number;
	/** Subcutaneous insulin in compartment 1. */
	Isc1: number;
	/** Subcutaneous insulin in compartment 2. */
	Isc2: number;
	/** Glucose mass in intestine. */
	Qgut: number;
	/** Carbs in stomach, solid phase. */
	Qsto1: number;
	/** Carbs in stomach, liquid phase. */
	Qsto2: number;
	/** Insulin in the interstitial fluid. */
	X: number;
	/** Insulin delay compartment 2. */
	XL: number;
	/** Increased energy consumption estimated through heart rate. */
	Y: number;
	/** Exercise-induced changes in insulin sensitivity. */
	Z: number;
	/** W. */
	W: number;
};

/**
 * Represents parameters for a UVA simulation.
 */
export type UvaPatientType = {
	/** Body weight. */
	BW: number;
	/** Steady-state of glucose in plasma. */
	Gpeq: number;
	/** Heart rate at rest. */
	HRb: number;
	/** Maximal heart rate (220-age). */
	HRmax: number; //220-age
	/** Distribution volume of glucose. */
	VG: number;
	/** Rate parameter from Gp to Gt. */
	k1: number;
	/** Rate parameter from Gt to Gp. */
	k2: number;
	/** Distribution volume of insulin. */
	VI: number;
	/** Rate parameter from Il to Ip. */
	m1: number;
	/** Rate parameter from Ip to Il. */
	m2: number;
	/** Rate parameter from Ip to periphery. */
	m4: number;
	/** Rate parameter of hepatic extraction (slope). */
	m5: number;
	/** Rate parameter of hepatic extraction (offset). */
	m6: number;
	/** Steady-state hepatic extraction of insulin. */
	HEeq: number;
	/** Maximal emptying rate of stomach. */
	kmax: number;
	/** Minimal emptying rate of stomach. */
	kmin: number;
	/** Rate constant of intestinal absorption. */
	kabs: number;
	/** Rate of grinding. */
	kgri: number;
	/** Fraction of intestinal absorption. */
	f: number;
	/** Extrapolated at zero glucose and insulin. */
	kp1: number;
	/** Liver glucose effectiveness. */
	kp2: number;
	/** Amplitude of insulin action on the liver. */
	kp3: number;
	/** Amplitude of portal insulin action on the liver. */
	kp4: number;
	/** Delay between insulin signal and insulin action. */
	ki: number;
	/** Glucose uptake by the brain and erythrocytes. */
	Fcns: number;
	/** Michaelis-Menten constant (offset). */
	Vm0: number;
	/** Michaelis-Menten constant (slope). */
	Vmx: number;
	/** Michaelis-Menten constant (offset). */
	Km0: number;
	/** Insulin action on the peripheral glucose utilization. */
	p2u: number;
	/** Glomerular filtration rate. */
	ke1: number;
	/** Renal threshold of glucose. */
	ke2: number;
	/** Rate constant of nonmonomeric insulin absorption. */
	ka1: number; // [Dalla Man, JDST, 2007]
	/** Rate constant of monomeric insulin absorption. */
	ka2: number; // [Dalla Man, JDST, 2007]
	/** Rate constant of insulin dissociation. */
	kd: number;
	/** Factor for exercise-induced increase in insulin sensitivity. */
	A: number; // [Dalla Man, JDST, 2009]
	/** Factor for exercise-induced increase in insulin-independent glucose clearance. */
	beta: number; // [Dalla Man, JDST, 2009]
	/** Factor for exercise-induced increase in glucose uptake. */
	gamma: number; // [Dalla Man, JDST, 2009]
	/** Parameter for calculating Z. */
	a: number; // [Dalla Man, JDST, 2009]
	/** Time constant for Y. */
	Thr: number; // [Dalla Man, JDST, 2009]
	/** Time constant for Z. */
	Tin: number; // [Dalla Man, JDST, 2009]
	/** Time constant for Z. */
	Tex: number; // [Dalla Man, JDST, 2009]
	/** Parameter for calculating Z. */
	n: number; // [Dalla Man, JDST, 2009]
};
