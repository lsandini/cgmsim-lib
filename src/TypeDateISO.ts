type TYear = `${number}${number}${number}${number}`;
type TMonth = `${number}${number}`;
type TDay = `${number}${number}`;
type THours = `${number}${number}`;
type TMinutes = `${number}${number}`;
type TSeconds = `${number}${number}`;
type TMilliseconds = `${number}${number}${number}`;

/**
 * Represent a string like `2021-01-08`
 */
type TDateISODate = `${TYear}-${TMonth}-${TDay}`;

/**
 * Represent a string like `14:42:34.678`
 */
type TDateISOTime = `${THours}:${TMinutes}:${TSeconds}.${TMilliseconds}`;

/**
 * Represent a string like `2021-01-08T14:42:34.678Z` (format: ISO 8601).
 */

export type TypeDateISO = `${TDateISODate}T${TDateISOTime}Z`;
