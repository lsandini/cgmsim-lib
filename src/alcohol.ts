import { TreatmentDelta, Treatment, TreatmentDrug, GenderType } from './Types';
import { getDrugActivity } from './drug';
import logger, { getDeltaMinutes, getTreatmentActivity } from './utils';

function getAlcoholActivity(
	gender: 'Male' | 'Female',
	weightKg: number,
	peakMin: number,
	durationMin: number,
	timeMin: number, //minutesAgo
	units: number,
): number {
	const peakAlcoholAmountMin = 60; //min
	const r = gender === 'Male' ? 0.68 : 0.55;
	const eliminationRate = gender === 'Male' ? 0.016 / 60 : 0.018 / 60; //g/100ml/min
	const peakAmount = ((units * 12) / (weightKg * 1000 * r)) * 100; //g/100ml
	const washoutDuration = peakAlcoholAmountMin + peakAmount / eliminationRate; //min
	const unitsWeighted = units * 12 * (80 / weightKg);
	if (washoutDuration < timeMin) {
		return (
			getTreatmentActivity(
				peakMin,
				durationMin,
				timeMin - washoutDuration,
				unitsWeighted,
			) / 35
		);
	}
	return 0;
}

export const computeAlcoholActivity = (
	treatments: TreatmentDelta[],
	weight: number,
	gender: GenderType,
) => {
	const treatmentsActivity = treatments.map((e) => {
		const minutesAgo = e.minutesAgo;
		const units = e.units;
		const activity = getAlcoholActivity(
			gender,
			weight,
			e.peak,
			e.duration,
			minutesAgo,
			units,
		);
		return activity;
	});
	logger.debug('these are the last Alcohol: %o', treatmentsActivity);
	const resultAct = treatmentsActivity.reduce((tot, activity) => {
		return tot + activity;
	}, 0);
	return resultAct;
};

export default function (
	treatments: TreatmentDrug[],
	weight: number,
	gender: GenderType,
): number {
	//Find Alcohol boluses

	const lastALC = getDrugActivity(treatments, weight, 'ALC');
	const activityALC =
		lastALC.length > 0 ? computeAlcoholActivity(lastALC, weight, gender) : 0;
	logger.debug('these are the last ALC: %o', { lastALC, activityALC });

	return activityALC / 2;
}
