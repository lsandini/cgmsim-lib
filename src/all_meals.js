const moment = require('moment');
module.exports = function(treatments = []) {
  const now = moment();
  const meals = treatments
    .filter(e => e.carbs && moment(e.mills).diff(now, 'minutes') > -360)
    .map(e => ({
      ...e,
      time:  (Date.now() - moment(e.mills).valueOf()) / (1000 * 60)
    }));

  console.log('Last 6 hours meal:', meals);
  return meals;
};
