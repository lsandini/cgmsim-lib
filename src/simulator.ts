import logger from './utils';

//const logger = pino();

import perlinRun from './perlin';
import computeBolusIRun from './computeBolusIOB.js';
import computeBasalIOBRun from './computeBasalIOB.js';
import allMealsRun from './all_meals.js';
import carbsRun from './carbs.js';
import arrowsRun from './arrows.js';
import liverRun from './liver.js';
import sgvStartRun from './sgv_start.js';
import { MainParams } from './Types';

logger.info('Run Init');

const main = ({
	env,
	entries,
	treatments,
	profiles,
	pumpBasals
}: MainParams) => {
	const weight = parseInt(env.WEIGHT);
	const dia = parseInt(env.DIA);
	const tp = parseInt(env.TP);
	const carbsAbs = parseInt(env.CARBS_ABS_TIME);
	const isf = parseInt(env.ISF);
	const cr = parseInt(env.CR);
	const perls = perlinRun(env.SEED || 'cgmsim');


	const bolusActivity = computeBolusIRun(treatments, dia, tp);
	const basalActivity = computeBasalIOBRun(treatments,weight);
	const lastMeals = allMealsRun(treatments);

	// //activity calc insulin
	// const det = detemirRun(weight, lastDET);
	// const gla = glargineRun(weight, lastGLA);
	// const degludec = degludecRun(lastDEG);
	// const tou = toujeoRun(weight, lastTOU);

	//activity calc carb
	const carbs = carbsRun(carbsAbs, lastMeals);
	const liver = liverRun(isf, cr);

	const cgmsim = sgvStartRun(entries, { basalActivity, liver, carbs,bolusActivity }, perls, isf);

	logger.info('this is the new sgv: %o', cgmsim);
	const arrows = arrowsRun([cgmsim, ...entries]);

	return { ...cgmsim, direction: arrows[0].direction };
};

export default main;