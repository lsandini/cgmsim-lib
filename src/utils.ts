import fetch from 'node-fetch';
import * as moment from 'moment';
import pino, { LevelWithSilent, TransportTargetOptions } from 'pino';
import setupParams from './setupParams';
import { Activity, Entry, EntryValueType, Note, Sgv, SimulationResult, TreatmentExpParam } from './Types';
import { load } from 'ts-dotenv';
import pinoPretty from 'pino-pretty';
import { TypeDateISO } from './TypeDateISO';

const env = load({
	LOGTAIL_SECRET: { type: String, optional: true },
	LOG_LEVEL: { type: String, optional: true },
	NODE_ENV: { type: String, optional: true },
});

const token: string = env.LOGTAIL_SECRET;
const level: LevelWithSilent | string = env.LOG_LEVEL ?? 'error';

const targets: TransportTargetOptions[] = [];

if (token) {
	targets.push({
		target: '@logtail/pino',
		options: { sourceToken: token },

		level,
	});
}

const logger = pino({
	level,
	prettifier: process.env.NODE_ENV === 'development' ? pinoPretty : null,
	transport: {
		targets,
	},
});

export default logger;

/**
 * Checks if a URL starts with HTTPS protocol
 * @param url - URL to check
 * @returns boolean indicating if the URL uses HTTPS
 */
export function isHttps(url: string | null | undefined): boolean {
	if (!url) {
		return false;
	}
	const httpsPattern = /^https:\/\//i;
	return httpsPattern.test(url);
}

/**
 * Removes trailing slash from a string if present
 * @param inputString - String to process
 * @returns String without trailing slash
 */
export function removeTrailingSlash(inputString: string): string {
	return inputString.endsWith('/') ? inputString.slice(0, -1) : inputString;
}

/**
 * Calculates exponential treatment activity based on given parameters
 * @param params - Treatment parameters including peak, duration, minutesAgo, and units
 * @returns Calculated activity value
 */
export function getExpTreatmentActivity({ peak, duration, minutesAgo, units }: TreatmentExpParam): number {
	const tau = (peak * (1 - peak / duration)) / (1 - (2 * peak) / duration);
	const scaleFactor = (2 * tau) / duration;
	const normalizationFactor = 1 / (1 - scaleFactor + (1 + scaleFactor) * Math.exp(-duration / tau));

	let activity =
		units *
		(normalizationFactor / Math.pow(tau, 2)) *
		minutesAgo *
		(1 - minutesAgo / duration) *
		Math.exp(-minutesAgo / tau);

	if (activity <= 0) {
		return 0;
	}

	// Ramp up activity linearly in first 15 minutes
	if (minutesAgo < 15) {
		return activity * (minutesAgo / 15);
	}

	return activity;
}
/**
 * Calculates exponential treatment activity based on given parameters
 * @param params - Treatment parameters including peak, duration, minutesAgo, and units
 * @returns Calculated activity value
 */
export function getExpTreatmentOnBody({ peak, duration, minutesAgo, units }: TreatmentExpParam): number {
	const tau = (peak * (1 - peak / duration)) / (1 - (2 * peak) / duration);
	const scaleFactor = (2 * tau) / duration;
	const normalizationFactor = 1 / (1 - scaleFactor + (1 + scaleFactor) * Math.exp(-duration / tau));

	let activity =
		units *
		(normalizationFactor / Math.pow(tau, 2)) *
		minutesAgo *
		(1 - minutesAgo / duration) *
		Math.exp(-minutesAgo / tau);

	if (activity <= 0) {
		return 0;
	}

	// Ramp up activity linearly in first 15 minutes
	if (minutesAgo < 15) {
		return activity * (minutesAgo / 15);
	}

	return activity;
}
const DEFAULT_SETTINGS = {
	KP: 3.0,
	KI: 0.3,
	KD: 1.8,
	TDI: 60,
	TARGET: 72, // mg/dL (6.0 mmol/L)
};

export const calculatePID = (
	entries: Sgv[],
	{ KP, KI, KD, TDI }: { KP: number; KI: number; KD: number; TDI: number } = DEFAULT_SETTINGS,
) => {
	try {
		if (!entries || entries.length < 36) {
			console.log('Entries received:', {
				entriesProvided: entries?.length || 0,
				required: 36,
				entries: entries,
			});
			throw new Error('Insufficient CGM data');
		}

		console.log('Starting PID calculation with entries:', entries.length);

		const settings = {
			target: DEFAULT_SETTINGS.TARGET,
			tdi: TDI,
			Kp: KP,
			Ki: KI,
			Kd: KD,
			maxBasalRate: (TDI / 24) * 1.5,
		};
		console.log('PID settings:', settings);

		const currentGlucose = entries[0].sgv;
		const error = (settings.target - currentGlucose) / 100;

		if (currentGlucose < settings.target) {
			console.log('Below target glucose - suspending insulin');
			return {
				rate: 0,
				diagnostics: {
					pTerm: 0,
					iTerm: 0,
					dTerm: 0,
					currentGlucose,
					suspendedForSafety: true,
				},
			};
		}

		const pTerm = -settings.Kp * error;

		const recentReadings = entries.slice(0, 24);
		const integralError = recentReadings.reduce((sum, reading, index) => {
			if (index === 0) return sum;
			const prevReading = recentReadings[index - 1];
			const timeGap = (prevReading.mills - reading.mills) / (1000 * 60);
			const prevError = (settings.target - prevReading.sgv) / 100;
			const currentError = (settings.target - reading.sgv) / 100;
			const avgError = (prevError + currentError) / 2;
			return sum + (avgError * timeGap) / 60;
		}, 0);

		const iTerm = -settings.Ki * integralError;

		let dTerm = 0;
		if (entries.length >= 3) {
			const window = entries.slice(0, 3);
			const timeSpanHours = (window[0].mills - window[2].mills) / (1000 * 60 * 60);
			const recentError = (settings.target - window[0].sgv) / 100;
			const olderError = (settings.target - window[2].sgv) / 100;
			const errorChange = recentError - olderError;
			const errorRateOfChange = errorChange / timeSpanHours;
			dTerm = -settings.Kd * errorRateOfChange;
		}

		let totalRate = Math.max(0, Math.min(pTerm + iTerm + dTerm, settings.maxBasalRate));
		const rate = Math.round(totalRate * 20) / 20;

		console.log('PID calculation result:', {
			rate,
			diagnostics: {
				pTerm,
				iTerm,
				dTerm,
				currentGlucose,
				suspendedForSafety: false,
			},
		});
		return {
			rate,
			diagnostics: { pTerm, iTerm, dTerm, currentGlucose },
		};
	} catch (error) {
		logger.error(`PID calculation error: %o`, error);
		throw error;
	}
};

/**
 * Calculates time difference in minutes between now and given timestamp
 * @param timestamp - Timestamp in milliseconds or ISO string
 * @returns Number of minutes difference
 */
export const getDeltaMinutes = (timestamp: number | TypeDateISO): number =>
	Math.round(moment().diff(moment(timestamp), 'seconds') / 60);

/**
 * Uploads data to Nightscout API
 * @param data - Data to upload (Entry, Activity, Note, or SimulationResult)
 * @param apiUrl - Nightscout API URL
 * @param apiSecret - API secret key
 * @returns Promise<void>
 */
export function uploadBase(
	data: Entry | Activity | Note | SimulationResult,
	apiUrl: string,
	apiSecret: string,
): Promise<void> {
	const isSecure = isHttps(apiUrl);
	const { postParams } = setupParams(apiSecret, isSecure);
	const jsonData = JSON.stringify(data);

	return fetch(apiUrl, {
		...postParams,
		body: jsonData,
	})
		.then(() => {
			logger.debug('[utils] Successfully updated Nightscout');
		})
		.catch((error) => {
			logger.debug('[utils] %o', error);
			throw new Error(error);
		});
}

/**
 * Loads data from Nightscout API
 * @param apiUrl - Nightscout API URL
 * @param apiSecret - API secret key
 * @returns Promise with array of entries
 */
export function loadBase(apiUrl: string, apiSecret: string): Promise<(Entry | Activity | Note)[]> {
	const isSecure = isHttps(apiUrl);
	const { getParams } = setupParams(apiSecret, isSecure);

	return fetch(apiUrl, {
		...getParams,
	})
		.then((response) => {
			logger.debug('[utils] Successfully loaded from Nightscout');
			return response.json();
		})
		.catch((error) => {
			logger.debug('[utils] %o', error);
			throw new Error(error);
		});
}

/**
 * Rounds a number to 8 decimal places
 * @param value - Number to round
 * @returns Rounded number
 */
export function roundTo8Decimals(value: number): number {
	const multiplier = Math.pow(10, 8);
	return Math.round(value * multiplier) / multiplier;
}
