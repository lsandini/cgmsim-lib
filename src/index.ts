import simulator from './CGMSIMsimulator';
import simulatorUVA from './UVAsimulator';
import downloads from './downloads';
import uploadEntries from './upload-entries';
import uploadActivity from './upload-activity';
import loadActivity from './load-activity';
import uploadNotes from './upload-notes';
import arrows from './arrows';


export { arrows, simulator, downloads, uploadEntries, uploadActivity, loadActivity, simulatorUVA, uploadNotes };
