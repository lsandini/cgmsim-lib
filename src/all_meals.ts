import logger from './utils';
import { Treatment, TreatmentDelta } from './Types';
import { getDeltaMinutes } from './utils';

//const logger = pino();

export default function (treatments: Treatment[]): TreatmentDelta[] {

	const meals = treatments
		.filter(e => e.carbs && getDeltaMinutes(e.mills) > 360)
		.map(e => ({
			...e,
			minutesAgo: getDeltaMinutes(e.mills),
		}));

	logger.info('Last 6 hours meal:', meals);
	return meals;
}

