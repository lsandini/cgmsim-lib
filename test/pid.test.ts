import { PatientInfoCgmsim, Sgv, NSTreatment, Activity, MainParams } from '../src/Types';
import simulator from '../src/CGMSIMsimulator';
import moment = require('moment');
import { diffOptions, getPngSnapshot, testGenerator } from './inputTest';
import { TypeDateISO } from '../src/TypeDateISO';
import { calculatePID } from '../src/utils';
const { toMatchImageSnapshot } = require('jest-image-snapshot');

const math = global.Math;

describe('pid test', () => {
  let date;
  const nsUrl = 'testUser';

  beforeEach(() => {
    date = new Date('2022-05-29T22:10:00Z');
    expect.extend({ toMatchImageSnapshot });

    jest.useFakeTimers('modern');
    jest.setSystemTime(date);
    const mockMath = Object.create(global.Math);
    mockMath.random = () => 0.5;
    global.Math = mockMath;
  });

  afterAll(() => {
    jest.useRealTimers();
    global.Math = math;
  });

  test('start from 180 @13:00Z', async () => {
    let now = moment('2022-06-04T13:00:00.000Z');
    jest.setSystemTime(now.toDate());
    const startSgv = 180;
    const entries: Sgv[] = [
      {
        mills: now.add(-5, 'minutes').toDate().getTime(),
        sgv: startSgv,
      }
    ];
    now = moment('2022-06-04T13:00:00.000Z');
    const patient: PatientInfoCgmsim = {
      CARBS_ABS_TIME: 360,
      CR: 10,
      DIA: 3,
      ISF: 36,
      TP: 55,
      WEIGHT: 250 / 3,
      AGE: 51,
      GENDER: 'Male',
      TZ: 'UTC',
    };

    let lastSgv = 0;
    const pumpEnabled = true;
    const sgvS = [];
    const TempBasalTreatment = [];
    for (let index = 0; index < 60 * 24; ) {
      const result = simulator({
        patient: patient,
        entries,
        treatments: [...TempBasalTreatment],
        profiles: [
          {
            startDate: '2000-01-01T00:00:00.000Z',
            defaultProfile: 'Default',
            store: {
              Default: {
                basal: 0.7,
              },
            },
          },
        ],
        pumpEnabled,
        user: { nsUrl },
      });

      entries.unshift({
        mills: now.valueOf(),
        sgv: result.sgv,
      });
      let resultPID;
      if (entries.length > 36) {
        resultPID = calculatePID(entries, { KP: 2.5, KI: 0.5, KD: 0.7, TDI: 60 });
        TempBasalTreatment.push({
          eventType: 'Temp Basal',
          rate: resultPID.rate,
          absolute: resultPID.rate,
          duration: 5,
          durationInMilliseconds: 5 * (60 * 1000),
          created_at: moment(now).toISOString(),
        });
      }

      for (let i = 0; i < 5; i++) {
        sgvS.push(result.sgv);
      }

      now = now.add(5, 'minutes');
      index = index + 5;
      jest.setSystemTime(now.toDate());
      lastSgv = result.sgv;
    }
    expect(lastSgv).toMatchSnapshot();
    const png = await getPngSnapshot(
      {
        type: 'multiple',
        values: [sgvS.map((sgv, index) => ({ key: index, value: sgv }))],
      },
      { scaleY: 400 },
    );

    expect(png).toMatchImageSnapshot(diffOptions);
    return;
  });
});
