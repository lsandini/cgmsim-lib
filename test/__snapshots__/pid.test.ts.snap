// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pid test start from 180 @13:00Z 1`] = `103`;
